File: Data Structures & Algorithms/trapping-rain-water/submission-13.java
class Solution {
    public int trap(int[] height) {
        int i=0,j=height.length-1;
        int maxL=height[i],maxR=height[j];
        int res=0;
        while(i<j){
            if(height[i]<=height[j]){
                maxL=Math.max(maxL,height[i]);
                res+=maxL-height[i];
                i++;
            }
            else {
                maxR=Math.max(maxR,height[j]);
                res+=maxR-height[j];
                j--;
            }
        }
        return res;
    }
}

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-13.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        i,j=0,0
        l=set()
        count=0
        while j<len(s):
            while s[j] in l: 
                l.remove(s[i])
                i+=1
            l.add(s[j])
            count=max(count,j-i+1)
            j+=1
        return count

